sailfishusermanager.cpp

```diff
@@ -348,3 +370,98 @@ bool SailfishUserManager::hasAccessRights(uint uid_to_modify) {
 
     return true;
 }
+
+void SailfishUserManager::switchUser(uint uid)
+{
+    if (checkCallerUid() == SAILFISH_UNDEFINED_UID)
+        return;
+
+    if (m_switchUser) {
+        auto message = QStringLiteral("Already switching user");
+        qCWarning(lcSUM) << message;
+        sendErrorReply(QDBusError::Failed, message);
+        return;
+    }
+
+    bool uidFound = false;
+    struct group *grent = getgrnam(USER_GROUP);
+    if (grent) {
+        for (int i = 0; grent->gr_mem[i]; i++) {
+            struct passwd *pw = getpwnam(grent->gr_mem[i]);
+            if (pw && pw->pw_uid == uid) {
+                uidFound = true;
+                break;
+            }
+        }
+    }
+    if (!uidFound) {
+        auto message = QStringLiteral("User not found");
+        qCWarning(lcSUM) << message;
+        sendErrorReply(QDBusError::InvalidArgs, message);
+        return;
+    }
+
+    if (sd_seat_get_active("seat0", nullptr, &m_currentUid) < 0) {
+        auto message = QStringLiteral("Failed to get current user id");
+        qCWarning(lcSUM) << message;
+        sendErrorReply(QDBusError::Failed, message);
+        return;
+    }
+
+    m_switchUser = uid;
+    m_timer->start();
+
+    QDBusInterface systemd(SYSTEMD_MANAGER_DEST, SYSTEMD_MANAGER_PATH, SYSTEMD_MANAGER_INTERFACE, QDBusConnection::systemBus());
```

Comment:
It feels a bit odd that this object is created separately for each call to the interface. It could be a member that is created when used first time. It would exist only if switchUser is used and would consume just a (null) pointer of memory otherwise.

Reply:
Reusing interface now

Reply:
It definitely makes sense to make it global or something. Or even combine it with D-Bus call wrappers as a single entity responsible for D-Bus communication with systemd.